### DISCUSSION TITLE
fabrikt fails to generate classes with API docs with "allOf" schemas

### DISCUSSION THREAD

**Discussion by Christian1984** [cjbooms/fabrikt#326] [Q&A]
Hi! I am trying to generate an API Client and models from an openapi v3.0.1 API documentation using fabrikt v17.4.0. 

The document in question can be found here: https://developers.shippeo.com/catalog/api/37ef941f-f328-34c4-acbf-02d7dcb032fc/doc?page=8fcf2df0-00e2-417d-8f2d-f000e2e17d79 (json-File can be downloaded with the Download button) - The documentation makes excessive use of `"allOf"` tags, which I assume are the root cause of my problem.

I am using the follwing excerpt from my `pom.xml` to create my Kotlin Sources with Maven:

```xml
<!-- Exec Maven Plugin to run Fabrikt -->
<plugin>
	<groupId>org.codehaus.mojo</groupId>
	<artifactId>exec-maven-plugin</artifactId>
	<version>3.0.0</version>
	<executions>
		<execution>
			<id>generate-pairing-api</id>
			<goals>
				<goal>java</goal>
			</goals>
			<configuration>
				<arguments>
					<argument>--api-file</argument>
					<argument>${project.basedir}/openapi/shippeo/pairing-api.json</argument>
					<argument>--base-package</argument>
					<argument>com.example.client.shippeo.api.pairing</argument>
					<argument>--http-client-target</argument>
					<argument>OK_HTTP</argument>
					<argument>--targets</argument>
					<argument>HTTP_MODELS</argument>
					<argument>--targets</argument>
					<argument>CLIENT</argument>
					<argument>--validation-library</argument>
					<argument>JAKARTA_VALIDATION</argument>
				</arguments>
			</configuration>
		</execution>
	</executions>
	<configuration>
		<mainClass>com.cjbooms.fabrikt.cli.CodeGen</mainClass>
	</configuration>
</plugin>
```

and then run `mvn exec:java@generate-pairing-api -X`. The build works fine, even without any warnings, but the code generator generates code that references missing classes and thus won't compile.

For example, `SharedSituationJustificationUncharterCharter.kt` looks like this

```kotlin
package com.example.client.shippeo.api.pairing.models

import com.fasterxml.jackson.`annotation`.JsonProperty
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull

public data class SharedSituationJustificationUncharterCharter(
  @param:JsonProperty("agency")
  @get:JsonProperty("agency")
  @get:NotNull
  @get:Valid
  public val agency: CharterAgency, // 👈 this class is not present and thus won't compile
)
```

In `SharedSituationJustificationUnpairPair.kt` a class `PairMean` is referenced which does not exist!

Is this a problem with the provided api documentation, my configuration or are there issues with fabrikt that cause this?

**Comment by cjbooms** [comment]
Most issues normally relate to deep nested incline object definitions, where fabrikt has to guess the name of the classes to generate. Especially so when those inclined definitions occur within polymorphic operators such as allof. 
Extracting some of the inline definitions to first-class definitions using ref usually helps.

**Comment by Christian1984** [comment]
Thanks for getting back to me! I understand, but then why does it guess the class name and reference it without creating it or a proper interface? If this happens, wouldn‘t you expect some warning or, if the code does not even build, expect an error?Am 15.10.2024 um 20:44 schrieb Conor Gallagher ***@***.***>:﻿
Most issues normally relate to deep nested incline object definitions, where fabrikt has to guess the name of the classes to generate. Especially so when those inclined definitions occur within polymorphic operators such as allof.
Extracting some of the inline definitions to first-class definitions using ref usually helps.

**Reply by cjbooms** [reply]
If you can boil it down to a minimum example spec in yaml that fails we can create an issue to take a look.
